### PR TITLE
Allow build override in PR from existing image

### PR DESCRIPTION
**Pull request by arjkesh**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description

PRs can get cumbersome when we just want to iterate on tests, but need to build an image from scratch. Some builds, like the one tested in this PR, take around 2.5 hours to build.

This PR introduces a build_override flag - only to be used in PRs - which allows you to override builds with a different image. This brings wait time from ~2.5 hours down to ~5 minutes (example [2549939](https://github.com/aws/deep-learning-containers/pull/3735/commits/2549939fa2cb5d82b440708c82d18d834c3a8b5e))

NOTE:

- tried to override autopatch, but autopatch is done at the buildspec level and not at the image level, and autopatch flags are scattered around, so for now this only works if you disable autopatch in your buildspec

### Tests run
- Tested two repo workflows in ([2549939](https://github.com/aws/deep-learning-containers/pull/3735/commits/2549939fa2cb5d82b440708c82d18d834c3a8b5e))
- Confirm autopatch still wins if enabled :/ ([703c319](https://github.com/aws/deep-learning-containers/pull/3735/commits/703c3193831291edc5fb298fcf8a06936ece9681))

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [x] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**
<details>
<summary>Expand</summary>

- [ ] `sagemaker_remote_tests = true`
- [ ] `sagemaker_efa_tests = true`
- [ ] `sagemaker_rc_tests = true`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

</details>

### Formatting
- [x] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

#### Builds to Execute
<details>
<summary>Expand</summary>

Click the checkbox to enable a build to execute upon merge.

*Note: By default, pipelines are set to "latest". Replace with major.minor framework version if you do not want "latest".*

- [ ] build_pytorch_training_latest
- [ ] build_pytorch_inference_latest
- [ ] build_tensorflow_training_latest
- [ ] build_tensorflow_inference_latest

</details>

### PR Checklist 
<details>
<summary>Expand</summary>

- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

#### NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ec2_benchmark_tests = true` or `sagemaker_benchmark_tests = true`
</details>

### Pytest Marker Checklist
<details>
<summary>Expand</summary>

- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type
</details>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/image_builder.py [hunk 1]
```diff
@@ -305,10 +303,19 @@ def image_builder(buildspec, image_types=[], device_types=[]):
             "buildspec_path": buildspec,
         }
 
+        beta_tag_override = image_config.get("beta_tag_override")
+        dockerfile = image_config("docker_file")
+        if beta_tag_override and not is_autopatch_build_enabled(buildspec_path=buildspec):
+            with tempfile.NamedTemporaryFile(delete=False) as temp_file_handle:
+                temp_file_handle.write(
+                    f"FROM {os.getenv('ACCOUNT_ID')}.dkr.ecr.{os.getenv('REGION')}.amazonaws.com/{image_repo_uri.replace('pr-', 'beta-')}:{beta_tag_override}"
```
Comment (arjkesh) [inline src/image_builder.py:311]: this replace logic wouldn't work for some cases - review after POC

File: src/config.py [hunk 2]
```diff
@@ -56,16 +56,13 @@ def is_autopatch_build_enabled_in_image_buildspec(buildspec_path=None):
...
     image_buildspec_object.load(buildspec_path)
     autopatch_build_flag = image_buildspec_object.get("autopatch_build", "False").lower() == "true"
-    print(f"Here: {buildspec_path}")
     return autopatch_build_flag
 
 
 def is_autopatch_build_enabled(buildspec_path=None):
-    return (
-        parse_dlc_developer_configs("build", "autopatch_build")
-        or os.getenv("AUTOPATCH")
```
Comment (arjkesh) [inline src/config.py:66]: I don't think this is needed anymore, @shantanutrip let me know if it is needed

File: src/config.py [hunk 3]
```diff
@@ -56,16 +56,13 @@ def is_autopatch_build_enabled_in_image_buildspec(buildspec_path=None):
     image_buildspec_object = Buildspec()
     image_buildspec_object.load(buildspec_path)
     autopatch_build_flag = image_buildspec_object.get("autopatch_build", "False").lower() == "true"
-    print(f"Here: {buildspec_path}")
```
Comment (arjkesh) [inline src/config.py:59]: remove debug

File: src/config.py [hunk 4]
```diff
@@ -56,16 +56,13 @@ def is_autopatch_build_enabled_in_image_buildspec(buildspec_path=None):
...
     return autopatch_build_flag
 
 
 def is_autopatch_build_enabled(buildspec_path=None):
-    return (
-        parse_dlc_developer_configs("build", "autopatch_build")
-        or os.getenv("AUTOPATCH")
-        or is_autopatch_build_enabled_in_image_buildspec(buildspec_path=buildspec_path)
-    )
+    return parse_dlc_developer_configs(
```
Comment (arjkesh) [inline src/config.py:63]: might want to remove this dev toml option as well, if is_autopatch_build_enabled_in_buildspec is already set to true. A bit confusing to have tow controllers for this behavior

Comment (shantanutrip) [reply src/config.py:63]: Yes, if possible, lets remove the autopatch_build flag from the dev toml and the condition over here.

File: src/image_builder.py [hunk 5]
```diff
@@ -211,10 +209,34 @@ def image_builder(buildspec, image_types=[], device_types=[]):
...
+                    f"TAG OVERRIDE MUST BE OF FORMAT {tag_override_regex}, but got {tag_override}. Proceeding with regular build."
+                )
+            else:
+                repo_override, t_override = tag_override.split(":")
+                with tempfile.NamedTemporaryFile(mode="w", delete=False) as temp_file_handle:
+                    source_uri = (
+                        f"{image_repo_uri.replace('pr-', f'{repo_override}-')}:{t_override}"
+                    )
+                    temp_file_handle.write(
+                        f"FROM {source_uri}\nLABEL dlc.dev.source_uri={source_uri}"
```
Comment (shantanutrip) [inline src/image_builder.py:230]: Why do we need this label?

Comment (shantanutrip) [reply src/image_builder.py:230]: Technically, we are making a new replica of the beta image, with just the Label as the difference, right? Would it change the SHA of the image and treat it as a new image? What if we have a beta image and a test fails. We the want to run the test for that beta image on the PR and to study it and tweak the test to get a success on the PR. Thereafter, we want to skip the test on the pipeline to allow the image to go to next stage since the test has succeeded on PR. In that case, is it possible that we unknowingly move the tag of the the original beta image to a new image via our PR? We might want to change this behavior if that happens.

Comment (shantanutrip) [reply src/image_builder.py:230]: I guess our tagging mechanism of the new image still uses the same logic as the previous one, so I guess we would not run through this situation. However, just wanted to confirm.

Comment (arjkesh) [reply src/image_builder.py:230]: Yes, we are basically ferrying over a beta or pr image to a new PR tag, but adding a label to reference the source image. We do not modify existing beta tags in this process - this option is only enable-able in PR context

File: src/image_builder.py [hunk 6]
```diff
@@ -140,10 +141,7 @@ def image_builder(buildspec, image_types=[], device_types=[]):
         if is_nightly_build_context():
             additional_image_tags.append(tag_image_with_date(image_tag))
 
-        if build_context != "PR":
-            image_tag = tag_image_with_datetime(image_tag)
-        # If build is not enabled, we don't care about the datetime tag
-        elif is_build_enabled():
+        elif is_build_enabled() or build_context != "PR":
```
[the remark this answers is not in the document]
Comment (arjkesh) [reply src/image_builder.py:144]: elif is pointing to nightly context. I suppose we can tag nightly images without datetime tag as well with no harm, I can update this
